components/scream/src/physics/register_physics.hpp

```diff
@@ -13,8 +13,6 @@
 #include "physics/cld_fraction/atmosphere_cld_fraction.hpp"
 #include "physics/rrtmgp/atmosphere_radiation.hpp"
```

Comment:
This is actually a general comment about the `register_physics.hpp` file.  If one were to use it, does that require that the executable also link to every process registered?  In other words, for a shoc+cld+p3 test we would still need to link to the rrtmgp library in CMakeLists.txt because register_physics would be looking for the RRTMGP information?

If so, how do we plan to handle things when we expand the number of processes to include say more than one microphysics scheme or similar.  Would we still need to include all of the known libraries in each test's CMake file?

Reply:
Including the file by itself does not add linking requirements. It is ok to declare functions that are never defined or linked, so long as they are never invoked (though it is bad coding).

However, if one were to actually call `register_physics()`, then all the physics libraries would have to be linked. That means that single-parametrization AD test (or, in general, tests that do not use all parametrizations) should not call `register_physics()`, or else the test would need to link against all physics parametrizations (which causes unneeded build deps).

The comment in the file hints at how one could still use `register_physics()` for single-parametrization tests: guard each include (and `register_product` call) by something like `#ifdef HAS_P3`, and add `HAS_P3` to the compile definitions of each lib. That way, when the lib is "linked" (*) via `target_link_libraries`, is `HAS_BLAH` are propagated to the test, and its include/register sections are compiled.

(*) recall that `target_link_libraries(foo PUBLIC bar)` does more than just linking `bar` to `foo`: it propagates public compiler flags/definitions from `bar` to `foo`.